### ISSUE TITLE
Missing initial WAL segment during restore

### ISSUE THREAD

**Issue by benbjohnson** [benbjohnson/litestream#224]
Posted from the Discussions page:

I just tried restoring a database from S3 and ran into an interesting error message:

```
cannot find max wal index for restore: missing initial wal segment: generation=4f2abd0f421cf473 index=00001c73 offset=1080
```

I was able to restore from the same generation yesterday so I'm wondering what this message means and what might have messed up to get into this state.

I can grab any listings of the contents of the S3 which might help if it sounds like a bug rather than something I've done.

_Originally posted by @pfw in https://github.com/benbjohnson/litestream/discussions/223_

**Comment by benbjohnson** [comment]
@pfw Thanks for reporting the bug. I moved it from GitHub Discussions to an issue to track it better. If you can provide a listing of the S3 bucket for `generations/4f2abd0f421cf473/wal` then that would help a lot. I suspect that the restore could have occurred while retention enforcement was running which may have removed the WAL files behind the scenes.

A couple of questions:

1. Do you have a timestamp of when you run the command?
2. Do you see any log messages from the Litestream process around that same time?
3. What operating system (Linux, macOS, etc) and CPU architecture (amd64 or ARM) are you running?
4. Can you provide your config settings (if you're using a config file)? Obviously, remove any sensitive info first. :)

**Comment by pfw** [comment]
Some more interesting bits... 

That generation is not present any more on s3. I've checked and there aren't any lifecycle rules on the bucket that might be deleting the generation.

At the moment I don't have any logs on the litestream replicate process - I'll make sure I capture those. Is it worth setting up with 'trace' or is that way too much data to run for days?

1. No timestamp specified when I first got the error but I did try with one and the result was the same.
2. No log messages
3. Ubuntu 18.04.5 LTS (GNU/Linux 5.4.0-1046-gcp x86_64)
4.  Only sync interval being set to 1m above a very basic s3 config
```
access-key-id: XYZ
secret-access-key: !@#

dbs:
  - path: DB
    replicas:
      - url: s3://bucket/DB
        sync-interval: 1m

```
5. Built from source litestream-0.3.5

**Comment by pfw** [comment]
A little extra - it was over 36 hours on that generation, the db it is replicating is very much only written to during business hours.

**Comment by benbjohnson** [comment]
> _That generation is not present any more on s3. I've checked and there aren't any lifecycle rules on the bucket that might be deleting the generation._

Hmm, that definitely makes debugging more difficult. A new generation can be started if Litestream is restarted and another process checkpoints the WAL file. Lifecycle rules on the database should maintain the same replica.

> _At the moment I don't have any logs on the litestream replicate process - I'll make sure I capture those._

Thanks! That would be helpful to see why the generation restarted.

> _ Is it worth setting up with 'trace' or is that way too much data to run for days?_

That's probably too much data to be running for days. It mainly traces the copy from the WAL file to the staging copy of the WAL files and wouldn't tell us too much about the replica state. The `litestream replicate` logs should say when retention is enforced so that would show when old WAL files are deleted. That occurs every 24 hours by default.

A couple more questions:
1. How long does it take to run a `litestream restore` for your database?
2. Unrelated by why do you build from source rather than use the release binaries? Nothing wrong with that—just genuinely curious.

**Comment by pfw** [comment]
Some _good_ news here...

I was trying to restore the db to see how long it takes and ran into the same error:

```
cannot find max wal index for restore: missing initial wal segment: generation=068c96cca219db86 index=000003f7 offset=1080
```

I've grabbed the output of snapshots, generations and wal and also listings of the s3 bucket.

[litestream-restore-error.txt](https://github.com/benbjohnson/litestream/files/6787755/litestream-restore-error.txt)

1. I restarted the replicate so I could restore and it took about 1min which is pretty much just the transfer time.
2. I don't actually recall the original reason for building from source, I suspect it was due to some issue running the prebuilt binary and that was a couple of releases ago. I just tried with the current prebuilt binary and that appears to work fine so I'll go with that from now on.

I now have the replicate logs being stored somewhere I can get at easily if I can capture this error again.

As an aside this restore process is for a development environment on my local machine (Mac although the error is the same when run from the machine doing the replicate) and I'd vote for any kind of incremental restore, at the moment I'm thinking I'll mirror the s3 bucket locally and have a config to restore from files as a way to be able to pick up changes during the day, sounds like it should work but not ideal.

**Comment by benbjohnson** [comment]
Thanks for dumping out the WAL files. It looks like there's definitely a missing initial WAL segment, however, I'm not really sure why. The only place in the code that collects WAL files for deletion is here and it filters out anything below a given index:

https://github.com/benbjohnson/litestream/blob/08d6adc660d0eefb28ea8b3305dbdc783e9aadd9/replica.go#L601-L607

The only other option I could think of is the WAL segment failing to write which would be logged in the `litestream replicate` logs. 

Can you try reverting to `v0.3.4` and seeing if it still happens? You may need to delete the `.db-litestream` file next to your database to clear it out and give it a fresh start. Also, I think the reason you had to build from source is that the Linux release version was built against a newer version of GLIBC which was fixed in https://github.com/benbjohnson/litestream/issues/198. You'll probably need to build from source for v0.3.4.

**Comment by pfw** [comment]
Thanks. I've reverted back to v0.3.4 and will try to restore regularly to see if it happens again.

**Comment by pfw** [comment]
I've just checked and got the restore error again on v0.3.4...

This restore is on a Mac machine
```

~/go/bin/litestream restore -v -config litestream.yml -o DB DB
/Users/pfw/DB(s3): restoring snapshot 21859587b250f593/0000005a to DB.tmp
/Users/pfw/DB(s3): restoring wal files: generation=21859587b250f593 index=[0000005a,0000026e]
cannot download wal 21859587b250f593/0000005a: out of sequence wal segments: 21859587b250f593/0000005a at remote offset 1080, expected offset 0
```

The replicate logs from the production linux machine
```
initialized db: /DB
replicating to: name="s3" type="s3" bucket="spatialgroupcomau-backups" path="property_daily.db" region="" endpoint="" sync-interval=1m0s
replicating to: name="s3" type="s3" bucket="spatialgroupcomau-backups" path="property_daily.db" region="" endpoint="" sync-interval=1m0s
/DB: sync: new generation "21859587b250f593", no generation exists
/DB(s3): snapshot: creating 21859587b250f593/00000000 t=5.923s
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 4b2acddca08cce4d/0000019c n=284
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/000000f0 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 4b2acddca08cce4d/0000019c n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/000000f0 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 4b2acddca08cce4d/0000019c n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/000000f0 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 4b2acddca08cce4d/0000019c n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/-0000001 n=3
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 4b2acddca08cce4d/-0000001 n=320
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000000 n=0
/DB(s3): snapshot: creating 21859587b250f593/0000005a t=5.854s
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/-0000001 n=1
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/-0000001 n=6
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/-0000001 n=9
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/-0000001 n=5
/DB(s3): retainer: deleting wal files before 21859587b250f593/0000005a n=0
/DB(s3): snapshot: creating 21859587b250f593/00000275 t=6.628s
/DB(s3): retainer: deleting wal files before 21859587b250f593/00000275 n=0
/DB(s3): retainer: deleting wal files before db24ae436bb549ba/-0000001 n=1

```

The bucket contents:

```
❯ aws s3 ls  s3://BUCKET/DB/generations/                           
                           PRE 21859587b250f593/

❯ aws s3 ls  s3://BUCKET/DB/generations/21859587b250f593/snapshots/
2021-07-12 15:17:48  239120763 00000275.snapshot.lz4

❯ aws s3 ls  s3://BUCKET/DB/generations/21859587b250f593/wal/
2021-07-12 15:19:47       1827 00000275_00000438.wal.lz4
2021-07-12 15:20:47        492 00000276_00000000.wal.lz4
```

**Comment by benbjohnson** [comment]
Hmm, that's really odd. I haven't seen that behavior before and it's strange that it happens so consistently for you. It looks like you're using 1KB pages—is that correct?

**Comment by pfw** [comment]
Yes, it is a 1k page. This database has had a long life... the oldest records are dated 2003 I can't be sure then the database file itself would have been created fresh (from a full dump and reload) so it is potentially 10+ years old. I'll do a dump and restore with current Sqlite defaults and then see if I get the same errors. It's nice that Sqlite can keep functioning with old files so it's has just moved between production machines over that time.

**Comment by benbjohnson** [comment]
Wow, that’s an old SQLite file.

The 1KB page size _shouldn’t_ be an issue but I would be curious if you see the issue when it’s loaded into a SQLite database with recent defaults.

I’m adding additional logging to the next version to show when WAL segments are written and deleted. I’m really not sure why it’s deleting that initial file after creating a new snapshot.

**Comment by pfw** [comment]
I'll get that new version of the db sorted as soon as I can. Let me know if there is a test version of litestream you'd like me to use. We have other backup processes in place while trying out litestream so I can easily put something in place for testing.

**Comment by pfw** [comment]
Hi Ben. I've found the issue and it was driver error, many apologies for the goose chase.

What was happening was there was a misconfiguration on a separate machine that ended up writing into the same bucket location. It looks like the delete on one DB/generation was hitting the other.

I wonder if this bit of the docs should be a bit stronger or there is something that could be done to detect this state and fail on the replicate?

> This approach also has the benefit that two servers that accidentally share the same replica destination will not overwrite each other’s data. However, note that it is not recommended to replicate two databases to the same exact replica path.

**Comment by benbjohnson** [comment]
Thanks for taking the time to track that down.

I agree that having two databases writing to the same bucket is a problem but I'm not entirely sure how to detect it in a good way. One option would be to use a lock file on the bucket that essentially holds a lease. However, there's a trade-off with how often that lease needs to be renewed versus how quickly a node can be recovered. For example, if a node updates the timestamp of the file every minute to continue to hold the lease then there could be up to a minute where a node could not recover. There's also a small additional cost for those `PUT` requests.

**Comment by pfw** [comment]
Agreed, that anything to deal with this user error could easily become over engineering.

Something simple that might work could be to have the ```replicate``` create a directory with the name of the db in the destination so that even if the same bucket(/path) is used it is only if two DBs with the same source path would clash, eg.

Rather than having something like this:
```
dbs:
  - path: /var/lib/db
    replicas:
      - url: s3://mybkt.litestream.io/db
```
the replicate url would be like:

```
      - url: s3://mybkt.litestream.io/
```

And the ```replicate``` would create a directory for the replica as ```base name /var/lib/db```.

Much more than something like that would just make setup more difficult.

**Comment by amadeojimenez** [comment]
I am having the same issue. My setup:

I am using litestream ( version 0.3.8 ) to replicate 3 different databases from a Ec2 instance to S3 and then restore them in a second Ec2 instance.  My .yml looks:

` dbs:
 - path: /home/xxx/db1.db
   replicas:
   - url:  s3://replicasbucket/db1.db
 
 - path: /home/xxx/db2.db
   replicas:
   - url:  s3://replicasbucket/db2.db

 - path: /home/xxx/db3.db
   replicas:
   - url:  s3://replicasbucket/db3.db`

At the beginning it works great but after some days getting the error:

`cannot find max wal index for restore: missing initial wal segment: generation=1d56061ee8dfc385 index=00000011 offset=4152`

when trying to restore db2.db. I erased the corrupted folder in S3 and restarted litestream: It seemed to work but again three days later I found the same issue (this time with db3.db)

May be I misconfigured something? I am willing to send any content if this might help. 

Thanks!

**Comment by benbjohnson** [comment]
@amadeojimenez Are you backing up to the same S3 bucket + path from multiple servers? This issue can happen if you're doing something like a rolling restart. Unfortunately, Litestream only works with a single server.

**Comment by amadeojimenez** [comment]
> @amadeojimenez Are you backing up to the same S3 bucket + path from multiple servers? This issue can happen if you're doing something like a rolling restart. Unfortunately, Litestream only works with a single server.
1
Yes, this is indeed the case. I will restrict to the single server apps. Thanks four your help!

**Comment by hifi** [comment]
I know this is an old issue and there are multiple reasons why one may end up in this situation, however, I think I've found some real bugs that may cause this which I attempted to fix in #522 and #527 if one finds this issue later.

Because Litestream currently does a passive checkpoint during snapshots without forcing a write, it's possible to lose the zero offset of a snapshot if there are no writes to the database after.

The second is an external problem where the replica may have old unrelated WAL segments in due to partial retention cleanup and that will cause this error even though it doesn't need to be fatal.
